mesa/time.py

```diff
@@ -198,3 +198,78 @@ def step(self) -> None:
             self.time += self.stage_time
 
         self.steps += 1
+
+
+class RandomActivationByType(RandomActivation):
+    """
+    A scheduler which activates each type of agent once per step, in random
+    order, with the order reshuffled every step.
+
+    This is equivalent to the NetLogo 'ask breed...' and is generally the
+    default behavior for an ABM.
+
+    Assumes that all agents have a step() method.
+
+    This implementation assumes that the type of an agent doesn't change
+    throughout the simulation.
+    """
+
+    def __init__(self, model: Model) -> None:
+        super().__init__(model)
+        self.agents_by_type = defaultdict(dict)
+
+    def add(self, agent: Agent) -> None:
+        """
+        Add an Agent object to the schedule
+
+        Args:
+            agent: An Agent to be added to the schedule.
+        """
+
+        self._agents[agent.unique_id] = agent
+        agent_class: Type[Agent] = type(agent)
+        self.agents_by_type[agent_class][agent.unique_id] = agent
+
+    def remove(self, agent: Agent) -> None:
+        """
+        Remove all instances of a given agent from the schedule.
+        """
+
+        del self._agents[agent.unique_id]
+
+        agent_class: Type[Agent] = type(agent)
+        del self.agents_by_type[agent_class][agent.unique_id]
+
+    def step(self, by_type: bool = True) -> None:
```

Comment:
Can we add something like an optional `order` input that allows to input the order in which the agent types are executed? Maybe that input could be either the string `"reversed"` to loop through `reversed(self.agents_by_type)` or a custom list with agent types.

I suggested a code snipped below.

Reply:
That's a good idea. Also good to compare with the API of NetLogo, MASON, and Agents.jl.

Reply:
Agents.jl has 2 params: `shuffle_types`, `shuffle_agents`: https://github.com/JuliaDynamics/Agents.jl/blob/5d1182dc7dd80001492c27333898c50462bdb20e/src/submodules/schedulers.jl#L92-L99.

Reply:
NetLogo' leaves the order up to the user's implementation with `ask [breed]`: https://github.com/NetLogo/models/blob/2bfb84e92b367f312e397338e27a4c85889f13a0/Sample%20Models/Biology/Wolf%20Sheep%20Predation.nlogo#L57-L75.

Reply:
Maybe we can allow three strings for `order`: `regular` (default), `reverse` and `shuffle`, and a list of agent types.

Reply:
Can't find a wolf-sheep model for MASON. It is also N/A in https://juliadynamics.github.io/Agents.jl/stable/comparison/.